src/coreComponents/constitutive/relativePermeability/TableRelativePermeability.hpp

```diff
@@ -0,0 +1,260 @@
+{
+public:
+
+  TableRelativePermeability( std::string const & name, dataRepository::Group * const parent );
+
+  virtual ~TableRelativePermeability() override;
+
+  std::unique_ptr< ConstitutiveBase > deliverClone( string const & name, Group * const parent ) const override;
+
+  static std::string catalogName() { return "TableRelativePermeability"; }
+
+  virtual string getCatalogName() const override { return catalogName(); }
+
+  /**
+   * @brief Create all the table kernel wrappers needed for the simulation (for all the phases present)
+   */
+  void createAllTableKernelWrappers();
+
+  /// Type of kernel wrapper for in-kernel update
+  using KernelWrapper = TableRelativePermeabilityUpdate;
+
+  /**
+   * @brief Create an update kernel wrapper.
+   * @return the wrapper
+   */
+  KernelWrapper createKernelWrapper();
+
+  struct viewKeyStruct : RelativePermeabilityBase::viewKeyStruct
+  {
+    static constexpr auto phaseMinVolumeFractionString = "phaseMinVolumeFraction";
+    static constexpr auto waterOilRelPermTableNamesString = "waterOilRelPermTableNames";
+    static constexpr auto gasOilRelPermTableNamesString = "gasOilRelPermTableNames";
+
+    using ViewKey = dataRepository::ViewKey;
+    ViewKey phaseMinVolumeFraction = { phaseMinVolumeFractionString };
+    ViewKey waterOilRelPermTableNames = { waterOilRelPermTableNamesString };
+    ViewKey gasOilRelPermTableNames = { gasOilRelPermTableNamesString };
+
+  } vieKeysTableRelativePermeability;
+
+protected:
+
+  virtual void postProcessInput() override;
+
+  virtual void initializePreSubGroups( Group * const ) override;
+
+private:
+
+  /**
+   * @brief Validate the relative permeability table provided in input (increasing phase vol frac and rel perm, etc)
+   * @param[in] relPermTable the relative permeability table (kr vs s) for a given phase)
+   */
+  real64 validateRelativePermeabilityTable( TableFunction const & relPermTable ) const;
+
+  /// Relative permeability table names (one for each phase in the oil-water pair)
+  array1d< string > m_waterOilRelPermTableNames;
+
+  /// Relative permeability table names (one for each phase in the oil-gas pair)
+  array1d< string > m_gasOilRelPermTableNames;
+
+  /// Relative permeability kernel wrapper for two-phase oil water data
+  array1d< TableFunction::KernelWrapper > m_waterOilRelPermTableKernelWrappers;
+
+  /// Relative permeability kernel wrapper for two-phase oil gas data
+  array1d< TableFunction::KernelWrapper > m_gasOilRelPermTableKernelWrappers;
+
+  /// Min phase volume fractions (deduced from the tables). With Baker, only the water phase entry is used
+  array1d< real64 > m_phaseMinVolumeFraction;
+
+};
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void
+TableRelativePermeabilityUpdate::
+  compute( arraySlice1d< real64 const > const & phaseVolFraction,
+           arraySlice1d< real64 > const & phaseRelPerm,
+           arraySlice2d< real64 > const & dPhaseRelPerm_dPhaseVolFrac ) const
+{
+  localIndex const NP = numPhases();
```

Comment:
Probably `nPhases` instead of `NP` is less misleading (`NP` looks like a template parameter/constant)

Reply:
@TotoGaz yes I agree I am going to make the change in the relperm models.